libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m

```diff
@@ -624,34 +640,25 @@ - (void) webView:(WKWebView *) webView decidePolicyForNavigationAction:(WKNaviga
         NSString *refreshUrl = [self isLoginRedirectUrl:navigationAction.request.URL];
         if (refreshUrl != nil) {
             [SFSDKHybridLogger w:[self class] message:@"Caught login redirect from session timeout. Reauthenticating."];
-            
+
+            // Auth success callback block.
+            __weak typeof(self) weakSelf = self;
+            SFUserAccountManagerSuccessCallbackBlock authSuccessCallbackBlock = ^(SFOAuthInfo *authInfo, SFUserAccount *userAccount) {
+                __strong typeof(weakSelf) strongSelf = weakSelf;
+                [SFUserAccountManager sharedInstance].currentUser = userAccount;
+                [strongSelf authenticationCompletion:refreshUrl authInfo:authInfo];
+            };
+
             /*
              * Reconfigure user agent. Basically this ensures that Cordova whitelisting won't apply to the
              * WKWebView that hosts the login screen (important for SSO outside of Salesforce domains).
              */
-            __weak typeof(self) weakSelf = self;
             [SFSDKWebUtils configureUserAgent:[self sfHybridViewUserAgentString]];
-            [self refreshCredentials:[SFUserAccountManager sharedInstance].currentUser.credentials
-             completion:^(SFOAuthInfo *authInfo, SFUserAccount *userAccount) {
-                 [SFUserAccountManager sharedInstance].currentUser = userAccount;
-
-                 // Reset the user agent back to Cordova.
-                 [weakSelf authenticationCompletion:refreshUrl authInfo:authInfo];
-             } failure:^(SFOAuthInfo *authInfo, NSError *error) {
-                 __strong typeof(weakSelf) strongSelf = weakSelf;
-                 if ([strongSelf logoutOnInvalidCredentials:error]) {
-                     [SFSDKHybridLogger e:[strongSelf class] message:@"Could not refresh expired session. Logging out."];
-                     NSMutableDictionary *attributes = [[NSMutableDictionary alloc] init];
-                     attributes[@"errorCode"] = [NSNumber numberWithInteger:error.code];
-                     attributes[@"errorDescription"] = error.localizedDescription;
-                     [SFSDKEventBuilderHelper createAndStoreEvent:@"userLogout" userAccount:nil className:NSStringFromClass([strongSelf class]) attributes:attributes];
-                     [[SFUserAccountManager sharedInstance] logout];
-                 } else {
-                     
-                     // Error is not invalid credentials, or developer otherwise wants to handle it.
-                     [strongSelf loadErrorPageWithCode:error.code description:error.localizedDescription context:kErrorContextAuthExpiredSessionRefresh];
-                 }
-             }];
+            if (![SFUserAccountManager sharedInstance].currentUser) {
```

Comment:
This is the crux of the changes. If no user is present, call `loginWithCompletion`. If there is a user present, call `refreshCredentials` like before.